### PR TITLE
OpenMapTiles 3.15.0 SNAPSHOT (3)

### PR DESCRIPTION
**Pull request by phanecak-maptiler**
Follow-up on #144 , contains bugfixes for problems found while comparing tilesets generated with 3.15.0 SNAPSHOT of OpenMapTiles and `planetiler-openmaptiles` and also final bump of version from `3.14.0` to `3.15.0`:

- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/41
  - plus https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/42
  - plus https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/50
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/43
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/44
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/45
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/46
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/47
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/48
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/49
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/51
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/55
- https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/58

### CODE REVIEW

**Comment by phanecak-maptiler** [comment]
note: Once this is merged, it fixes #15 thanks to https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/48 .

**Review by lazaa32 — APPROVED** [review]

File: README.md
Comment (msbarry) [inline README.md:170]: Interesting to add the emphasis here, I thought the plan was to switch to CC0?

File: src/main/java/org/openmaptiles/layers/WaterName.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/WaterName.java:235]: typ: `bellow` -> `below`

File: src/main/java/org/openmaptiles/layers/WaterName.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/WaterName.java:237]: why? Is planetiler-openmaptiles handling considered "better" here?

File: src/main/java/org/openmaptiles/layers/WaterName.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/WaterName.java:250]: You should be able to replace this handling with a combination of min area and min zoom?

```java
feature.setMinPixelSizeBelowZoom(13, 128).setMinZoom(minzoom)
```

File: src/main/java/org/openmaptiles/layers/Water.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/Water.java:254]: Good point. I did it the same way as OpenMaptiles, but since "bigger area" implies "bigger ratio" we can drop `/ neGeom.getArea()`. Adjusted, thanks.

File: src/main/java/org/openmaptiles/layers/WaterName.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/WaterName.java:235]: Adjusted.

File: src/main/java/org/openmaptiles/layers/WaterName.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/WaterName.java:237]: Bay name is rendered along the centerline, instead of as usual label used for points.

File: src/main/java/org/openmaptiles/layers/WaterName.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/WaterName.java:250]: Will that work if we're creating a point from that polygon with `features.pointOnSurface()`?

Comment (msbarry) [reply src/main/java/org/openmaptiles/layers/WaterName.java]: Yes! I updated it to work that way in order to support use-cases like this here: https://github.com/onthegomap/planetiler/pull/720

File: src/main/java/org/openmaptiles/layers/Water.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:139]: I think we want to do an atomic operation on the concurrent hashmap here to avoid multiple updates at the same time from colliding with eachother:

```java
neLakeNameMap.merge(lakeInfo.name, (prev, next) -> next.geom.getArea() > prev.geom.getArea() ? next : prev);
```

File: src/main/java/org/openmaptiles/layers/Water.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:254]: do you still need this?

File: src/main/java/org/openmaptiles/layers/Water.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:246]: seems like there might be a risk here of multiple OSM threads updating the same lakeInfo instance at the same time, would probably be safer to wrap this in `synchronized (lakeInfo) { ... }`

File: src/main/java/org/openmaptiles/OpenMapTilesMain.java
Comment (msbarry) [reply src/main/java/org/openmaptiles/OpenMapTilesMain.java:4]: OK I see you fixed the roads at z4/5, but I'm still seeing these discrepancies in planetiler-openmaptiles:

- M74 doesn't have "e-road" network set: https://onthegomap.github.io/planetiler-demo/#17/55.687717/-3.942389 - did we decide to completely get rid of e-roads and AsianHighway designation after that omt pr?
- That Peru maritime boundary still shows up in planetiler-openmaptiles output at z4: https://onthegomap.github.io/planetiler-demo/#4.4/-9.01/-78.94

Those are showing latest output from main branch of planetiler-openmaptiles from @ZeLonewolf's setup

**Review by msbarry — COMMENTED** [review]
looks good!  a few more comments...

File: src/main/java/org/openmaptiles/layers/Water.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:189]: I see this is firing 16 times when run over the planet, but I can't tell which line is actually causing this failure?  Can you catch the TopologyException closer to where it happens so that 1) we get a more detailed error message that helps pinpoint where the issue happened and 2) we can skip the outer layer of try/catch(geometry exception) and just have the final attempt log the error like this: https://github.com/onthegomap/planetiler/blob/12558b99a8074781e2adc9a2130cab5f481b4395/planetiler-core/src/main/java/com/onthegomap/planetiler/geo/GeoUtils.java#L311-L338.  Also please pass `config.logJtsExceptions()` into the 4th argument of `e.log(...)` below which will print out more details about when the error happens?

I understand this is a best-effort attempt to add OSM IDs to low zoom lakes so it's fine if it misses some, we just try to track down any of these robustness issues with JTS pretty aggressively to report to its maintainer.

Here's an example failure for reference:

> 0:56:26 WAR [osm_pass2:process] - Unable to process intersections for OSM feature 13070460: error fixing polygon: org.locationtech.jts.geom.TopologyException: found non-noded intersection between LINESTRING ( 0.7828271086875376 0.3132446324915123, 0.7828058490561336 0.3132069744583109 ) and LINESTRING ( 0.7828058490561336 0.3132069744583109, 0.7828042136998719 0.3132040773818683 ) [(0.7828058490561336, 0.3132069744583109, NaN) ]; original error: org.locationtech.jts.geom.TopologyException: found non-noded intersection between LINESTRING ( 0.7828271086875376 0.3132446324915123, 0.7828058490561336 0.3132069744583109 ) and LINESTRING ( 0.7828058490561336 0.3132069744583109, 0.7828042136998719 0.3132040773818683 ) [ (0.7828058490561336, 0.3132069744583109, NaN) ]

File: src/main/java/org/openmaptiles/OpenMapTilesMain.java
Comment (msbarry) [reply src/main/java/org/openmaptiles/OpenMapTilesMain.java:4]: Looks like this is still an issue on this branch:
- e-road: https://protomaps.github.io/PMTiles/?url=https%3A%2F%2Fprotomaps.dev%2F%7Emsbarry%2Fomt-3.15-branch.pmtiles#map=17/55.687717/-3.942389
- Peru border: https://protomaps.github.io/PMTiles/?url=https%3A%2F%2Fprotomaps.dev%2F%7Emsbarry%2Fomt-3.15-branch.pmtiles#map=4.4/-9.01/-78.94

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Transportation.java:269]: should this be named `isTrunkWithNetworkForZ4` ? Or should we also run this check down below:

```java
        case FieldValues.CLASS_MOTORWAY ->
          isMotorwayForZ4(routeRelations) ||
          isMotorwayWithNetworkForZ4(routeRelations) ?
          MINZOOMS.getOrDefault(FieldValues.CLASS_MOTORWAY, Integer.MAX_VALUE) : 5;
```

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Transportation.java:283]: I notice that this logic leads to short stretches of road in the US getting included at z4 because they have class=motorway without a route network and they are disconnected with the rest of the road because we exclude class=motorway/network=us-highway until z5 - is that the intent? 

I think it's because this logic is upgrading those trunks to motorways:

```java
          if (z5trunk && isMotorwayWithNetworkForZ4(routeRelations)) {
            clazz = FieldValues.CLASS_MOTORWAY;
```

image


https://protomaps.github.io/PMTiles/?url=https%3A%2F%2Fprotomaps.dev%2F%7Emsbarry%2Fomt-3.15-branch.pmtiles#map=4.8/36.05/-84.16

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Transportation.java:303]: did we mean to remove these? If so we should remove them from the enum list or otherwise keep them here.

File: src/main/java/org/openmaptiles/OpenMapTilesMain.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/OpenMapTilesMain.java:4]: As per M74: That seems to be related to https://github.com/openmaptiles/openmaptiles/pull/1648 and https://github.com/openmaptiles/openmaptiles/pull/1649 . I've regenerated that portion of UK and `e-road` is there:

Screenshot from 2024-04-29 13-47-00

Hence I guess Brian's tiles were done with older snapshot.

Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/OpenMapTilesMain.java:4]: As per Peru: Fix/change is here, originally tracked here: https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/47 . Generated with current snapshot:

Screenshot from 2024-04-29 13-53-08

So again, maybe older snapshot used by Brian.

File: src/main/java/org/openmaptiles/layers/Water.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:253]: Maybe add a little bit more context, like "fixed invalid OSM water polygon {} when adding ID to natural earth {}"

File: src/main/java/org/openmaptiles/layers/Water.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:224]: Maybe add a little bit more context, like "fixed invalid OSM water polygon {} when looking up natural earth features"

File: src/main/java/org/openmaptiles/layers/Water.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:225]: can you overwrite geom here so that fillOsmIdIntoNeLake doesn't need to fix the geometry again?

**Review by msbarry — COMMENTED** [review]
looks good, a few tweaks make the intersection error logs a bit more self-explanatory...

File: src/main/java/org/openmaptiles/layers/Water.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:224]: also let's omit the exception here and on line 253 below - there's no sense digging into JTS exceptions unless we're sure the input is valid.

**Comment by phanecak-maptiler** [comment]
Regarding latest batch of "adjusted" comments from me: You can see that only now, since only now I've pushed and merged those changes. Sorry for the delay and confusion.

File: src/main/java/org/openmaptiles/layers/Landcover.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Landcover.java:2]: I notice you've got 2023 in a few of these headers

File: src/main/java/org/openmaptiles/layers/Water.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Water.java:205]: looks like you need to run `./scripts/format.sh`

**Review by msbarry — APPROVED** [review]
looks good!

File: src/main/java/org/openmaptiles/layers/Landcover.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/Landcover.java:2]: Ah, @TomPohys was asking about the same.

TL;DR: IANAL hence I tend to put there a year of "most recent substantial change". This particular file has such in 2023.

Longer: Looking through various IP advice and Open Source licensing FAQs, etc. there seems to be varying degree of advice, from "does not matter" to "really important", with various tactics about when and how to change that value. And since I'm not going to hire a layer (which is IMHO important if we truly need a legal advice about for example what "substantial change" means), then I outline my rationale:

1. Extreme position 1: Bump the copyright year every year even when no changes were done = IMHO totally bogus and IANAL but legally unenforceable.
2. Extreme position 1: Put there a year at the beginning and then never bump it, even if the file goes through substantial changes all the time = IMHO also wrong, but sort of "wronging myself" (e.g. possibly loosing protections which I'm entitled to, but hey, copyright protection is supposed to run even several years after my death, so ...).

Hence I opt to be "somewhere in between", e.g. bump the year when substantial changes occur.

File: src/main/java/org/openmaptiles/layers/Water.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/Water.java]: Thanks, adjusted.

**Comment by phanecak-maptiler** [comment]
Merged.

Release planned for Monday. E.g. if no serious showstopper is found until than, this will be 3.15.

**Pull request merged**
